src/vision/status_quo/alan_thinks_he_needs_async_locks.md

```diff
@@ -0,0 +1,121 @@
+
+Barbara knows that mutex guards are not `Send` because sending mutex guards to different threads is not a good idea. She suggests looking into async locks which can be held across await points because they are `Send`. Alan looks into the tokio documentation for more info and is easily able to move the use of the standard library's mutex to tokio's mutex. It compiles!
+
+Alan ships his code and it gets a lot of usage. After a while, Alan notices some potential performance issues. It seems his proxy handler does not have the throughput he would expect. Barbara, having newly joined his team, sits down with him to take a look at potential issues. Barbara is immediately worried by the fact that the lock is being held much longer than it needs to be. The lock only needs to be held while accessing the route and not during the entire duration of the downstream request.
+
+She suggests to Alan to switch to not holding the lock across the I/O operations. Alan first tries to do this by explicitly cloning the url and dropping the lock before the proxy request is made:
+
+```rust
+impl Proxy {
+  async fn handle(&self, key: String, request: Request) -> crate::Result<Response> {
+      let routes = self.state.lock().unwrap();
+      let route = routes.get(key).unwrap_or_else(crate::error::MissingRoute)?.clone();
+      drop(routes);
+      Ok(self.client.perform_request(route, request).await?)
+  }
+}
+```
+
+Unfortunately this does not seem to fix the issue - he gets the same error message as before. Confused Alan goes to Barbara for advice. She is also confused, and it takes several minutes of exploration before she comes to a solution that works - wrapping the mutex access in a block and implicitly dropping the mutex:
+
+```rust
+impl Proxy {
+  async fn handle(&self, key: String, request: Request) -> crate::Result<Response> {
+      let route = {
+        let routes = self.state.lock().unwrap();
+        routes.get(key).unwrap_or_else(crate::error::MissingRoute)?.clone()
+      };
+      Ok(self.client.perform_request(route, request).await?)
+  }
+}
+```
+
+This works! Barbara mentions she's unsure why explicitly dropping the mutex guard did not work, but they're both happy that the code compiles. Shipping to production shows a large increase in throughput. Alan is really excited about Rust, and wants to write more!
+
+Barbara continues her own journey of learning even more about async Rust. After some enlightening talks at the latest RustConf, she decides to revisit the code that she and Alan wrote together. She asks herself, is using an *async* lock the right thing to do? This lock should only be held for a very short amount of time. Yielding to the runtime is likely more expensive than just synchronously locking. She decides to switch back to synchronous locks and is surprised when her initial benchmarking shows improvements in the 99th percentile of requests.
```

Comment:
I don't understand this 99th percentile reference -- did things get better? I can't really tell :)

Reply:
Yes I can rephrase that. I'm not actually sure how much difference this would make. It should only impact the service under very high load as yielding back to a very busy runtime will likely be slower than just blocking inside the future (since the critical section is very small). A read-write lock would further help in this case but I don't want to add unnecessary detail to the story.